add Jaeger tracing

beyang opened the pull request:
Add Jaeger tracing to Zoekt. This replaces the existing `golang.org/x/net/trace.New` calls with calls to `github.com/sourcegraph/zoekt/trace.New`. The latter invokes `net/trace` as before, but also creates an opentracing span.

![image](https://user-images.githubusercontent.com/1646931/88866616-0f5e8480-d1c0-11ea-916e-c596b8fbc719.png)


There are two new packages:
* `github.com/sourcegraph/zoekt/trace`: mimics the `net/trace` API, but creates an opentracing span under the hood.
* `github.com/sourcegraph/zoekt/trace/ot`: a wrapper around the opentracing API that enables selective tracing toggled by a context item.

These two packages are very similar to `github.com/sourcegraph/sourcegraph/internal/trace` and `github.com/sourcegraph/sourcegraph/internal/ot`. For now, it seems like they are different enough that it isn't worth the effort of merging them (which would mean breaking these packages out into a separate repository). If we find that we are frequently making changes to one and having to reflect them over to the other, we can revisit down the road.

This PR has a sister PR in the main Sourcegraph repo (https://github.com/sourcegraph/sourcegraph/pull/12548), which should be merged immediately after this is merged and it is updated with the latest dep version.

~Note that this does not yet propagate span IDs across the RPC boundary between `frontend` and `zoekt`, so the `zoekt` spans are "dangling".~ SpanContext is now propagated across the RPC boundary, so the zoekt spans are properly "attached" in the trace.

Questions:
* @keegancsmith this updates the RPC options struct. Will this break RPC calls between `frontend` and `zoekt-webserver` if this is not deployed atomically with the corresponding update to `frontend`?

Merging this will help with the effort to add more repos to the Sourcegraph.com global index.

## Reviewer guide: :point_right: [Start here](https://github.com/sourcegraph/zoekt/pull/55/files#r463240093)

beyang committed:
add Jaeger tracing

beyang committed:
accept propagated SpanContext

beyang commented:
From @keegancsmith:

> I'm not exactly sure on the value of tracing each shard access (yet). Tracing in the server side sounds good. But it won't contribute that much over what we are currently tracing from the client side (in frontend). This is because Zoekt returns stats about what it did in its response. I also recently improved the logging of that a lot. For example see the logs in this trace: https://sourcegraph.com/-/debug/jaeger/trace/54aa2901efff10e3?uiFind=4a9e3ea58ec2c84b
> 
> What are your thoughts on what is currently done vs adding tracing into zoekt + slightly complicating the binary. If we do want to add jaeger support in, I think we should consider building the zoekt binary in sourcegraph so we get access to internal/zoekt and conf pkg.

beyang commented:
> What are your thoughts on what is currently done vs adding tracing into zoekt + slightly complicating the binary.

As someone who is less familiar with Zoekt internals, having the Jaeger tracing helps me understand what's going on vs. I'm not exactly sure how to interpret the list of stats that's logged in the frontend-side trace. It also shows if one particular index file took much longer to read (as in the case of the screenshot in the PR description), whereas I'm not sure how to reproduce this information from the aggregate stats in what's currently logged.

> If we do want to add jaeger support in, I think we should consider building the zoekt binary in sourcegraph so we get access to internal/zoekt and conf pkg.

That seems like a good idea. Would that just involve moving `cmd/zoekt-webserver` into the Sourcegraph main repo? It would be nice to have zoekt have access to the conf pkg... but I think the overarching concern I have at the moment is increasing the size of the sourcegraph.com global search index—and determining whether Zoekt is the right solution for scaling up to 1M repositories. If it's not, then the work we do moving it into the Sourcegraph main repo would be a bit of a waste. Therefore, I'd like to defer moving it into the main repo for the time being.

TL;DR: this would help me with understanding Zoekt behavior so we can continue increasing the size of the global repo search index on sourcegraph.com. Happy to rip all this out if it later turns out not to be useful.

keegancsmith commented:
Really sorry I did't get around to reviewing this properly today. I promise to tackle it Monday.

Without actually reading the code these are my thoughts:

1. The RPC connection is stateful. You can communicate jaeger configuration from the frontend that is attached to the connection. That should quite nicely work I think? Unless jaeger is better with a global tracer.
2. I saw mention of communicating sampling strategy. Is that important? Shouldn't it be as simple as if an RPC has a span attached, use that span.
3. "Will this break RPC calls between frontend and zoekt-webserver". No, gob is backwards compatible for "missing" fields. https://golang.org/pkg/encoding/gob/#hdr-Types_and_Values
4. My main concern is introducing a performance penalty. For example when searching a shard that has no trigram matches zoekt does _very_ little work. I would want to ensure jaeger doesn't do much if no one ever looks at the trace. For example `net/trace` is really good for that. The `lazyprintf` only ever evaluates a printf if someone actually looks at it.
5. Tracing per shard is not that useful for understanding performance problems at scale. To be honest I can imagine this amount of tracing data will just overload the trace and make it near impossible to understand (or even load) any real trace. I'd be much happier if we just traced at the shardedSearcher level rather than at the index level.
6. The packages you add should live under `internal`, or we should consider adding a new repo with our nice `internal/trace` pkg :)

uwedeportivo reviewed (commented):
LGTM and i love the navigation

i'll let keegan accept

keegancsmith reviewed (changes requested):
The main blocker to landing this is `net/trace` is fine with per shard logging, but `ot` is not. If we turn this on we won't be able to load a trace for a search on sourcegraph.com since it will search over 100k+ shards => 100k+ spans of noise.

I'm not exactly sure on the solution to this. I think it is likely making it possible at the client call site for `trace` to opt out of `ot`.

beyang commented:
@keegancsmith I meant to do more work on this PR today, but got sidetracked, so just wanted to post what my plans are, since I only partially responded to the review comments:
* Will remove the tracing for index shards, since I agree that has a chance of causing a decent perf hit.
* I am hesitant to factor out the `trace` and `ot` packages into a separate repository, because it feels a bit premature to generalize here. The tracing needs of the Sourcegraph main repo differ a bit from what we want in Zoekt. For instance, the behavior of "only record spans if the context already contains a parent span" is not needed in the main repo's `trace` package, which is toggled by the tracing mode package global variable (`all`, `none`, `selective`). I think I'd rather have a package specific to this repository that satisfies *its* need for now. The existing `trace` and `ot` packages in this repository can be simplified to meet that requirement (e.g., no need for the tracing mode package global variable copied over from the main repo, as you point out), which I will do. In other words, I will simplify these packages in this repository, but I think it is better *not* to factor out the tracing packages into a separate dependency for now.
* I think it's better to keep the command-line flag to turn on/off Jaeger. It doesn't add *that* much complexity and it may come in handy in case we need to turn Jaeger off for whatever reason on sourcegraph.com or on a customer instance.
* The other issues you pointed out I think are minor and I'll address/fix them.

^ Lemme know if you have any objections to the above plan of action.

keegancsmith commented:
> * Will remove the tracing for index shards, since I agree that has a chance of causing a decent perf hit.

Great.

> * I am hesitant to factor out the `trace` and `ot` packages into a separate repository, because it feels a bit premature to generalize here. The tracing needs of the Sourcegraph main repo differ a bit from what we want in Zoekt. For instance, the behavior of "only record spans if the context already contains a parent span" is not needed in the main repo's `trace` package, which is toggled by the tracing mode package global variable (`all`, `none`, `selective`). I think I'd rather have a package specific to this repository that satisfies _its_ need for now. The existing `trace` and `ot` packages in this repository can be simplified to meet that requirement (e.g., no need for the tracing mode package global variable copied over from the main repo, as you point out), which I will do. In other words, I will simplify these packages in this repository, but I think it is better _not_ to factor out the tracing packages into a separate dependency for now.

Sounds good!

> * I think it's better to keep the command-line flag to turn on/off Jaeger. It doesn't add _that_ much complexity and it may come in handy in case we need to turn Jaeger off for whatever reason on sourcegraph.com or on a customer instance.

Adjusting the command line for docker containers is a pain in the arse to be honest. Probably better to make it an environment variable.

beyang committed:
update jaeger initialization

beyang committed:
simplify tracing packages

beyang committed:
remove index shard tracing

beyang committed:
docstring

beyang commented:
@keegancsmith PTAL

* I moved the `ot` package into `trace`. After simplifying, there were only a few symbols remaining, so it made sense to merge.
* I would like to punt adding tracing to `traceAwareSearcher.List` to a subsequent PR to be done immediately after this is merged (because I'd like to get this deployed in the next day or so).
* Removed opentracing for index shards.
* Removed the Jaeger flag, now using an environment variable (`JAEGER_DISABLED`, which is interpreted as default true if unset).

If remaining comments are minor, I'd prefer to address them in a follow-up PR if that's alright.

beyang committed:
remove log line
